File: cmd/cashier/parse.go
package main

import (
	"encoding/json"
	"regexp"
	"strings"
)

// ── types ─────────────────────────────────────────────────────────────────

type Block struct {
	Type  string   // "p", "h", "code", "ul", "ol", "quote"
	Level int      // headings
	Text  string   // p, h, quote
	Items []string // ul, ol
	Lang  string   // code
	File  string   // code
	Body  string   // code text
}

type Section struct {
	Type  string
	Props map[string]any
	Body  string
}

// ── inline formatting ─────────────────────────────────────────────────────

var (
	reBacktick  = regexp.MustCompile("`([^`]+)`")
	reBold      = regexp.MustCompile(`\*\*([^*]+)\*\*`)
	reItalic    = regexp.MustCompile(`(^|[^*])\*([^*\n]+)\*`)
	reLink      = regexp.MustCompile(`\[([^\]]+)\]\(([^)]+)\)`)
	reHardBreak = regexp.MustCompile(`\\\n`)
)

func esc(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	return s
}

func inline(src string) string {
	s := esc(src)
	s = reBacktick.ReplaceAllString(s, `<code class="code-inline">$1</code>`)
	s = reBold.ReplaceAllString(s, `<strong>$1</strong>`)
	s = reItalic.ReplaceAllStringFunc(s, func(m string) string {
		sub := reItalic.FindStringSubmatch(m)
		if sub == nil {
			return m
		}
		return sub[1] + "<em>" + sub[2] + "</em>"
	})
	s = reLink.ReplaceAllString(s, `<a href="$2">$1</a>`)
	s = reHardBreak.ReplaceAllString(s, "<br>")
	return s
}

// ── block parser ──────────────────────────────────────────────────────────

var (
	reHeading = regexp.MustCompile(`^(#{1,6})\s+(.*)$`)
	reUL      = regexp.MustCompile(`^[-*]\s`)
	reOL      = regexp.MustCompile(`^\d+\.\s`)
	reQuote   = regexp.MustCompile(`^>\s?`)
)

func parseBlocks(md string) []Block {
	lines := strings.Split(md, "\n")
	var blocks []Block
	i := 0
	for i < len(lines) {
		line := lines[i]

		// blank line
		if strings.TrimSpace(line) == "" {
			i++
			continue
		}

		// code fence
		if strings.HasPrefix(line, "```") {
			header := strings.TrimPrefix(line, "```")
			parts := strings.SplitN(strings.TrimSpace(header), " ", 2)
			lang, file := "", ""
			if len(parts) > 0 {
				lang = strings.TrimSpace(parts[0])
			}
			if len(parts) > 1 {
				file = strings.TrimSpace(parts[1])
			}
			i++
			var body []string
			for i < len(lines) && !strings.HasPrefix(lines[i], "```") {
				body = append(body, lines[i])
				i++
			}
			i++ // closing fence
			blocks = append(blocks, Block{Type: "code", Lang: lang, File: file, Body: strings.Join(body, "\n")})
			continue
		}

		// heading
		if m := reHeading.FindStringSubmatch(line); m != nil {
			blocks = append(blocks, Block{Type: "h", Level: len(m[1]), Text: m[2]})
			i++
			continue
		}

		// blockquote
		if reQuote.MatchString(line) {
			var parts []string
			for i < len(lines) && reQuote.MatchString(lines[i]) {
				parts = append(parts, reQuote.ReplaceAllString(lines[i], ""))
				i++
			}
			blocks = append(blocks, Block{Type: "quote", Text: strings.Join(parts, " ")})
			continue
		}

		// unordered list
		if reUL.MatchString(line) {
			var items []string
			for i < len(lines) && reUL.MatchString(lines[i]) {
				items = append(items, reUL.ReplaceAllString(lines[i], ""))
				i++
			}
			blocks = append(blocks, Block{Type: "ul", Items: items})
			continue
		}

		// ordered list
		if reOL.MatchString(line) {
			var items []string
			for i < len(lines) && reOL.MatchString(lines[i]) {
				items = append(items, reOL.ReplaceAllString(lines[i], ""))
				i++
			}
			blocks = append(blocks, Block{Type: "ol", Items: items})
			continue
		}

		// paragraph — consume consecutive non-block lines
		var paraLines []string
		for i < len(lines) {
			l := lines[i]
			if strings.TrimSpace(l) == "" {
				break
			}
			if strings.HasPrefix(l, "```") || reHeading.MatchString(l) ||
				reQuote.MatchString(l) || reUL.MatchString(l) || reOL.MatchString(l) {
				break
			}
			paraLines = append(paraLines, l)
			i++
		}
		if len(paraLines) > 0 {
			blocks = append(blocks, Block{Type: "p", Text: strings.Join(paraLines, " ")})
		}
	}
	return blocks
}

// ── section parser ────────────────────────────────────────────────────────

var reSectionFence = regexp.MustCompile(`(?m)^:::([\w-]+)(?:\s+(\{[^\n]*\}))?\s*\n([\s\S]*?)\n:::[ \t]*(?:\n|$)`)

func parseSections(md string) (string, []Section) {
	var sections []Section
	pre := ""
	firstIdx := reSectionFence.FindStringIndex(md)
	if firstIdx != nil {
		pre = strings.TrimSpace(md[:firstIdx[0]])
	}
	matches := reSectionFence.FindAllStringSubmatchIndex(md, -1)
	allMatches := reSectionFence.FindAllStringSubmatch(md, -1)
	for idx, m := range allMatches {
		_ = matches[idx]
		sType := m[1]
		propsJSON := m[2]
		body := strings.TrimSpace(m[3])

		props := map[string]any{}
		if propsJSON != "" {
			_ = json.Unmarshal([]byte(propsJSON), &props)
		}
		sections = append(sections, Section{Type: sType, Props: props, Body: body})
	}
	return pre, sections
}

File: cashier/parse.mjs
// cashier/parse.mjs — tiny markdown (no deps)
// Handles: headings, paragraphs, unordered + ordered lists, blockquote,
// inline **bold**, *italic*, `code`, [text](url), and line breaks.

export const esc = (s) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');

export function inline(src) {
  let s = esc(src);
  // code
  s = s.replace(/`([^`]+)`/g, '<code class="code-inline">$1</code>');
  // bold **x**
  s = s.replace(/\*\*([^*]+)\*\*/g, '<strong>$1</strong>');
  // italic *x*  (avoid ** already consumed)
  s = s.replace(/(^|[^*])\*([^*\n]+)\*/g, '$1<em>$2</em>');
  // links
  s = s.replace(/\[([^\]]+)\]\(([^)]+)\)/g, '<a href="$2">$1</a>');
  // hard break: trailing \\
  s = s.replace(/\\\n/g, '<br>');
  return s;
}

// Parse a markdown chunk into a list of { type, ... } blocks.
export function parseBlocks(md) {
  const lines = md.replace(/\r\n/g, '\n').split('\n');
  const blocks = [];
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    if (!line.trim()) { i++; continue; }

    // fenced code block ```lang [filename] ... ```
    if (/^```/.test(line)) {
      const info = line.slice(3).trim();
      const spaceIdx = info.indexOf(' ');
      const lang = spaceIdx === -1 ? info : info.slice(0, spaceIdx);
      const file = spaceIdx === -1 ? '' : info.slice(spaceIdx + 1).trim();
      const buf = [];
      i++;
      while (i < lines.length && !/^```/.test(lines[i])) { buf.push(lines[i]); i++; }
      i++; // skip closing fence
      blocks.push({ type: 'code', lang, file, text: buf.join('\n') });
      continue;
    }

    // heading
    const h = line.match(/^(#{1,6})\s+(.*)$/);
    if (h) { blocks.push({ type: 'h', level: h[1].length, text: h[2] }); i++; continue; }

    // blockquote
    if (line.startsWith('>')) {
      const buf = [];
      while (i < lines.length && lines[i].startsWith('>')) {
        buf.push(lines[i].replace(/^>\s?/, '')); i++;
      }
      blocks.push({ type: 'quote', text: buf.join(' ') });
      continue;
    }

    // unordered list
    if (/^[-*]\s/.test(line)) {
      const items = [];
      while (i < lines.length && /^[-*]\s/.test(lines[i])) {
        items.push(lines[i].replace(/^[-*]\s+/, '')); i++;
      }
      blocks.push({ type: 'ul', items });
      continue;
    }

    // ordered list
    if (/^\d+\.\s/.test(line)) {
      const items = [];
      while (i < lines.length && /^\d+\.\s/.test(lines[i])) {
        items.push(lines[i].replace(/^\d+\.\s+/, '')); i++;
      }
      blocks.push({ type: 'ol', items });
      continue;
    }

    // paragraph (until blank line)
    const buf = [];
    while (i < lines.length && lines[i].trim() && !/^(```|#{1,6}\s|>|[-*]\s|\d+\.\s)/.test(lines[i])) {
      buf.push(lines[i]); i++;
    }
    blocks.push({ type: 'p', text: buf.join(' ') });
  }
  return blocks;
}

// Split a full document into a sequence of fenced section chunks.
//   :::hero { "label": "00 Hero" }
//   ...markdown...
//   :::
// Anything before the first fence becomes a doc-level header (title + meta).
export function parseSections(md) {
  const out = [];
  const re = /^:::([\w-]+)(?:\s+(\{[^\n]*\}))?\s*\n([\s\S]*?)\n:::\s*(?=\n|$)/gm;
  let lastIndex = 0;
  let pre = '';
  let m;
  while ((m = re.exec(md)) !== null) {
    if (m.index > lastIndex) pre += md.slice(lastIndex, m.index);
    let props = {};
    if (m[2]) { try { props = JSON.parse(m[2]); } catch (e) {} }
    out.push({ type: m[1], props, body: m[3] });
    lastIndex = re.lastIndex;
  }
  if (lastIndex < md.length) pre += md.slice(lastIndex);
  return { pre: pre.trim(), sections: out };
}
